src/mlpack/methods/ann/augmented/tasks/add_impl.hpp

```diff
@@ -0,0 +1,148 @@
+/**
+ * @file add_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of AddTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_ADD_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_ADD_IMPL_HPP
+
+#include "add.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+AddTask::AddTask(const size_t bitLen) : bitLen(bitLen)
+{
+  if (bitLen <= 0) {
+    std::ostringstream oss;
+    oss << "AddTask::AddTask(): binary length (" << bitLen << ") "
+        << "is not positive!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+}
+
+const void AddTask::Generate(arma::field<arma::mat>& input,
+                             arma::field<arma::mat>& labels,
+                             const size_t batchSize,
+                             bool fixedLength)
+{
+  arma::field<arma::vec> vecInput = arma::field<arma::colvec>(batchSize);
+  arma::field<arma::vec> vecLabels = arma::field<arma::colvec>(batchSize);
+  size_t sizeA = bitLen, sizeB = bitLen;
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    if (!fixedLength)
+    {
+      // Generate random uniform length from [2..maxLength].
+      sizeA = mlpack::math::RandInt(2, bitLen + 1);
+      sizeB = mlpack::math::RandInt(2, bitLen + 1);
+    }
+    // Construct sequence of the form
+    // (binary number with sizeA bits) + '+'
+    // + (binary number with sizeB bits).
+    vecInput(i) = arma::randi<arma::colvec>(
+        sizeA + sizeB + 1, arma::distr_param(0, 1));
+    // Insert special value for '+' delimiter.
+    vecInput(i).at(sizeA) = 0.5;
+
+    int valA = 0;
+    for (size_t k = 0; k < sizeA; ++k)
+    {
+      valA += static_cast<int>(vecInput(i).at(k)) << k;
+    }
+
+    int valB = 0;
+    for (size_t k = sizeA + 1; k < sizeA + 1 + sizeB; ++k)
+    {
+      valB += static_cast<int>(vecInput(i).at(k)) << (k - sizeA - 1);
+    }
+
+    int tot = valA + valB;
+    std::vector<int> binarySeq;
+    while (tot > 0)
+    {
+      binarySeq.push_back(tot & 1);
+      tot >>= 1;
+    }
+    if (binarySeq.empty())
+    {
+      if (valA + valB != 0) {
+        std::ostringstream oss;
+        oss << "AddTask::Generate(): output sequence is empty "
+            << "but the target sum is not 0 (=" << valA + valB << ")"
+            << std::endl;
+        throw std::domain_error(oss.str());
+      }
+      binarySeq.push_back(0);
+    }
+    size_t totLen = binarySeq.size();
+    vecLabels(i) = arma::colvec(totLen);
+    for (size_t j = 0; j < totLen; ++j)
```

Comment:
I think we can directly use ``binarySeq.size();``instead of ``totLen`` what you think?